mtags/src/main/scala/scala/meta/internal/pc/CompilerJobQueue.scala

```diff
@@ -0,0 +1,84 @@
+package scala.meta.internal.pc
+
+import java.util.concurrent.ThreadPoolExecutor
+import java.util.concurrent.PriorityBlockingQueue
+import java.{util => ju}
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.CancellationException
+
+/**
+ * A thread pool executor to execute jobs on a single thread in a last-in-first-out order.
+ *
+ * The last-in-first-out order is important because it's common for Metals
+ * users to send multiple completion/hover/signatureHelp requests in rapid
+ * succession. In these situations, we care most about responding to the latest
+ * request even if it comes at the expense of ignoring older requests.
+ *
+ * To restrict unsafe multi-threaded access to the presentation compiler we
+ * schedule jobs to run on a single thread. We use this executor instead of the
+ * presentation compiler thread (see MetalsGlobalThread) for the following reasons:
+ * - we limit the usage of sleep/notify/wait/synchronized primitives.
+ * - some blocking compiler APIs like `ask[T](op: () => T)` don't seem to work as advertised.
+ * - it's preferable to work on top of CompletableFuture[T] instead of the custom Response[T] from the compiler,
+ *   which is required to execute tasks on the presentation compiler thread:
+ *   - CompletableFuture[T] can be passed via Java-only reflection APIs in mtags-interfaces.
+ *   - CompletableFuture[T] can be returned to lsp4j, for non-blocking JSON-RPC request handling.
+ *   - CompletableFuture[T] can be converted to Scala Futures for easier composition.
+ */
+class CompilerJobQueue(val executor: ThreadPoolExecutor) {
+  override def toString(): String = s"CompilerJobQueue($executor)"
+  def shutdown(): Unit = executor.shutdown()
+  def submit(fn: () => Unit): Unit = {
```

Comment:
can it just be `=> Unit` ?

Reply:
It can be but I don't think it would make the code more readable. As a general rule of thumb I avoid by-name parameters until it's become painful to write the leading `() =>`. In this case, we only call `submit` once.

Reply:
It's also a couple times in tests, but it's ok with me.

Reply:
I personally think it's more readable with the leading `() =>` at call-sites. It's only for APIs that are used a lot like `Option[T].getOrElse(=>T)` where I think by-name is justified.